dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/hibernate/HibernateTrackedEntityStore.java

```diff
@@ -1445,4 +1445,33 @@ private String getOrgUnitNameByUid(String uid) {
   private boolean skipOwnershipCheck(TrackedEntityQueryParams params) {
     return params.getUser() != null && params.getUser().isSuper();
   }
+
+  // TODO This has to be removed once SettingKey.TrackedEntityInstanceMaxLimit is removed
+  private int resolveTrackedEntityMaxLimit() {
+    int deprecatedTeiMaxLimit =
+        systemSettingManager.getIntegerSetting(SettingKey.DEPRECATED_TRACKED_ENTITY_MAX_LIMIT);
+    int newTeiMaxLimit =
+        systemSettingManager.getIntegerSetting(SettingKey.TRACKED_ENTITY_MAX_LIMIT);
+
+    if (isSet(deprecatedTeiMaxLimit)
+        && isSet(newTeiMaxLimit)
+        && deprecatedTeiMaxLimit != newTeiMaxLimit) {
+      throw new IllegalQueryException(
+          String.format(
+              "Both keys %s and %s cannot be present at the same time",
+              SettingKey.TRACKED_ENTITY_MAX_LIMIT.getName(),
+              SettingKey.DEPRECATED_TRACKED_ENTITY_MAX_LIMIT.getName()));
+    }
+
+    // both settings are disabled
+    if (!isSet(deprecatedTeiMaxLimit) && !isSet(newTeiMaxLimit)) {
+      return -1;
+    }
+
+    return isSet(newTeiMaxLimit) ? newTeiMaxLimit : deprecatedTeiMaxLimit;
+  }
+
+  private boolean isSet(Integer i) {
+    return i > 0;
```

Comment:
Are the KeyTrackedEntityInstanceMaxLimit and KeyTrackedEntityMaxLimit not 5000 by default? In which case this would be true by default for both 🤔

Reply:
yes. 0 or negative means setting is disabled

Reply:
> 0 or negative means setting is disabled

I am not sure this will work 🤔 Maybe I just don't get it yet 😅 

https://github.com/dhis2/dhis2-core/blob/e37f087868595ae4f635fcdba0e62b90c29dbf2b/dhis-2/dhis-services/dhis-service-setting/src/main/java/org/hisp/dhis/setting/SystemSettingManager.java#L71-L79

If a user does not save any of those keys both are set to a default which is 5000 > 0. So both keys will be set according to your logic IMHO (unless I miss something).

Reply:
yes, both will be set to 5000. Which is fine. we use any one of them. 
if both are set and have different values then we have a conflict.